[SE-0322] Temporary uninitialized buffers

grynspan opened the pull request:
<!-- What's in this pull request? -->
These changes add the ability to temporarily allocate buffers preferentially on the stack.

Pitch thread: https://forums.swift.org/t/pitch-temporary-uninitialized-buffers/48954
Evolution PR: https://github.com/apple/swift-evolution/pull/1379
Evolution thread: https://forums.swift.org/t/se-0322-temporary-uninitialized-buffers/51848

<!-- If this pull request resolves any bugs in the Swift bug tracker, provide a link: -->
Resolves rdar://76198203.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

grynspan commented:
@swift-ci please smoke test

grynspan commented:
@swift-ci please smoke test

grynspan commented:
@swift-ci please clean test

grynspan commented:
@swift-ci please smoke test

zoecarver commented on lib/IRGen/GenBuiltin.cpp:
```diff
@@ -643,6 +643,23 @@ if (Builtin.ID == BuiltinValueKind::id) { \
+      return;
+    }
+
+    auto alloc = IGF.emitDynamicAlloca(IGF.IGM.Int8Ty, size, align, Alignment(16), "builtin-stackAlloc");
```
So if the align isn't a compile time constant, you'll default to 16 bytes?

I think we could maybe develop a better API here. What if we had two overloads: one that takes an alignment argument and one that doesn't. If the alignment argument is provided and isn't a compile time constant you could create an error. This also means you wouldn't have to create another overload of `emitDynamicAlloca`.

grynspan replied:
It is intentional that the parameters to the calling API might be determined at runtime. The selection of 16-byte alignment is intentional: it's the default alignment for heap-allocated memory in Swift as well. Of course, we might want to fall back to something else or actually honour the alignment request where possible—I don't know LLVM well enough to know how to go about building that though.

grynspan replied:
(This is a WIP PR I opened so I could run tests. It needs to be reviewed both in swift-evolution and by the core team before it can hope to be merged.)

grynspan replied:
FWIW I'm working on new codegen here that should act a bit smarter.

grynspan commented:
@swift-ci please smoke test

grynspan commented:
@swift-ci please smoke test

grynspan commented:
@swift-ci please smoke test

grynspan commented:
@swift-ci please smoke test

varungandhi-apple commented on lib/IRGen/GenOpaque.cpp:
```diff
@@ -629,6 +629,82 @@ void IRGenFunction::emitDeallocateDynamicAlloca(StackAddress address) {
+  StackAddress stackPointer;
+  auto nullPointer = llvm::Constant::getNullValue(returnPtrType);
+
+  // Check that this function type can emit a stack allocation safely.
+  if (isAsync() || isCoroutine()) {
+    // These types of functions don't actually stack-allocate, and support for
+    // them is not needed at this time.
+    return nullPointer;
+  }
```
Could you explain this behavior? Naively, I would expect that you should be able to use similar scratch-space like functionality in async functions too. There the allocation would be part of the async frame instead of the stack frame, but everything else should work as-is.

Maybe the primitive should be called "frameAlloc" or something so that it is more general?

grynspan replied:
Async functions don't appear to emit `alloca` calls for dynamically-sized ranges. Instead, they heap-allocate and later free it. Since the API I'm proposing is not `async`, it doesn't need to care about these cases.

varungandhi-apple replied:
Ah, right. I wrote this earlier but later realized that the size is not necessarily statically known.

varungandhi-apple commented:
Please add some test cases under `test/IRGen` as well as `test/stdlib`.

varungandhi-apple commented on lib/IRGen/GenOpaque.cpp:
```diff
@@ -629,6 +629,82 @@ void IRGenFunction::emitDeallocateDynamicAlloca(StackAddress address) {
+  // Check that the caller-supplied alignment is usable.
+  Alignment resolvedAlign;
+  if (auto alignConst = dyn_cast<llvm::ConstantInt>(align)) {
+    // Ensure the supplied alignment is itself a power of 2.
+    auto alignConstValue = alignConst->getSExtValue();
+    if (!llvm::isPowerOf2_64(alignConstValue)) {
+      alignConstValue = llvm::NextPowerOf2(alignConstValue);
+    }
+
+    if (alignConstValue > 0) {
+      resolvedAlign = Alignment(alignConstValue);
+
+    } else {
+      // The specified alignment is nonsensical (negative, zero, or too big.)
+      return nullPointer;
+    }
+
```
Maybe this could be simplified a little?

```
auto alignConstValue = alignConst->getSExtValue();
if (alignConstValue <= 0 /* optionally add too big check */) {
  return nullPointer;
}
resolvedAlign = Alignment(llvm::PowerOf2Ceil(alignConstValue)); 
```

grynspan replied:
> Maybe this could be simplified a little?
> 
> ```
> auto alignConstValue = alignConst->getSExtValue();
> if (alignConstValue <= 0 /* optionally add too big check */) {
>   return nullPointer;
> }
> resolvedAlign = Alignment(llvm::PowerOf2Ceil(alignConstValue)); 
> ```

`PowerOf2Ceil()` rounds to zero on overflow, so I don't think this is necessarily a correct change.

varungandhi-apple replied:
`NextPowerOf2` also returns `0` on overflow, so I'm not sure what's different? We should probably add an upper limit check to `if (alignConstValue <= 0 /* optionally add too big check */)` anyways.

grynspan replied:
Sorry, yes, I'm using `NextPowerOf2()` already—which returns `0` on overflow.

Oh, I see what you're saying, `PowerOf2Ceil()` won't turn e.g. `8` into `16`, right?

varungandhi-apple replied:
It shouldn't turn 8 into 16.

```
/// Returns the power of two which is greater than or equal to the given value.
/// Essentially, it is a ceil operation across the domain of powers of two.
inline uint64_t PowerOf2Ceil(uint64_t A) {
  if (!A)
    return 0;
  return NextPowerOf2(A - 1);
}
```

grynspan replied:
Alright.

varungandhi-apple commented on lib/IRGen/GenOpaque.cpp:
```diff
@@ -629,6 +629,82 @@ void IRGenFunction::emitDeallocateDynamicAlloca(StackAddress address) {
+  result->addIncoming(stackPointer.getAddressPointer(), smallEnoughBB);
+  result->addIncoming(nullPointer, tooBigBB);
+
+  // FIXME: Box in an Optional. CreateBitCast?
+  // result = boxSomehow(result);
```
For an optional, looks like `%i = ptrtoint i8* %p to i64` should be the right thing on 64-bit. I don't know if there is a better function here or if you should directly have a bit-cast...

grynspan replied:
LLVM vomits if I try to emit a `ptrtoint` here, so there's something extra happening that I don't know how to resolve.

varungandhi-apple replied:
The return type would also need to be changed; right now it's `i8 *`. Maybe that's the source of the issue?

grynspan replied:
I'm honestly not certain—I need somebody better-versed in this stuff than me to chime in. Some comments elsewhere in the stdlib make me think this is an unavoidable problem. For now, the hackery I'm doing is good enough for testing, though.

grynspan replied:
We discussed off of GitHub. No obvious solution is presenting itself at this time.

varungandhi-apple commented on lib/IRGen/GenOpaque.cpp:
```diff
@@ -629,6 +629,82 @@ void IRGenFunction::emitDeallocateDynamicAlloca(StackAddress address) {
+    } Builder.CreateBr(doneBB);
+
+    Builder.emitBlock(tooBigBB); {
+      // Here's where we might call out to the stdlib to ask for assistance.
+      // The stdlib could e.g. check thread conditions to determine if it's safe
+      // to issue an alloca, at which point we'd emit one.
```
Could you elaborate on what you mean here? I don't quite understand. Is this a bug-fix/optimization opportunity/something else?

grynspan replied:
It's an optimization opportunity—TBD depending on core team feedback. Might be "nothing."

varungandhi-apple commented on stdlib/public/core/UnsafeUninitializedPointer.swift:
```diff
@@ -0,0 +1,140 @@
+/// - Parameters:
+///   - byteCount: The number of bytes to temporarily allocate. `byteCount` must
+///     not be negative.
+///   - alignment: The alignment of the new, temporary region of allocated
+///     memory, in bytes.
```
I'm not sure what the right phrasing is, but we should have wording for "compile-time constant" here and in other places where that is a requirement.

grynspan replied:
Given that there is no requirement that the alignment be a compile-time constant, I don't believe such wording is needed.

varungandhi-apple replied:
Maybe I'm misunderstanding something... There's this call in the body:

```
Builtin.stackAlloc(byteCount._builtinWordValue, alignment._builtinWordValue)
```

For this call to work, `alignment._builtinWordValue` needs to be a compile-time constant according the documentation for `stackAlloc` in `Builtins.def`. Doesn't that imply that `alignment` also needs to be a compile-time constant?

grynspan replied:
I think it just implies my comment in Builtin.defs is leaking implementation details. :)

grynspan commented:
> Please add some test cases under `test/IRGen` as well as `test/stdlib`.

Yep, it's on my to-do list. :)

benrimmington commented on stdlib/public/core/UnsafeUninitializedPointer.swift:
```diff
@@ -0,0 +1,140 @@
+  _debugPrecondition(byteCount > 0, "Too little uninitialized memory requested.")
+  _debugPrecondition(alignment > 0, "Nonsensical alignment requested.")
+
+  // FIXME: Builtin.stackAlloc() should return an Optional<Builtin.RawPointer>.
+  let stackPointer = Builtin.stackAlloc(byteCount._builtinWordValue, alignment._builtinWordValue)
+  if Int(Builtin.ptrtoint_Word(stackPointer)) != 0 {
+    defer { _fixLifetime(stackPointer) }
+    return try body(.init(start: UnsafeMutableRawPointer(stackPointer), count: byteCount))
```
There's a comment in [Runtime.swift](https://github.com/apple/swift/blob/5f41f94f42c9ddd56a18ae1e6c685c7fa5bb32b4/stdlib/public/core/Runtime.swift#L30):

> We use Builtin.Word here because Builtin.RawPointer can't be nil.

```suggestion
  let stackPtrOrNil: Builtin.Word = Builtin.stackAlloc(
    byteCount._builtinWordValue, alignment._builtinWordValue)
  if let start = UnsafeMutableRawPointer(bitPattern: Int(stackPtrOrNil)) {
    return try body(.init(start: start, count: byteCount))
```

Why is the `_fixLifetime` call needed?

grynspan replied:
I am aware of that comment, thanks! I am waiting for feedback from the core team regarding the correct way to express a nullable pointer in a builtin. :)

> Why is the _fixLifetime call needed?

To prevent the compiler from reclaiming the stack space occupied by the temporary allocation. See other uses of this idiom in the stdlib. (It's possible it's not needed here, but again am waiting for core team feedback.)

grynspan replied:
The returned value is of type `Builtin.RawPointer`. The last use of that value is _before_ the call to `body()` returns, because it's passed to `UnsafeMutableRawPointer.init(_:)`. Under strict lifetime rules, it is then eligible for destruction before `body()` is even called.

grynspan replied:
@benrimmington I attempted your proposed change here but it doesn't actually compile, so I'm backing it out. I'm going to wait for feedback from the core team to see how best to implement this pattern. :)

zoecarver reviewed (commented):
I'll submit these for now, but I have a few others to come.

grynspan commented on lib/IRGen/IRGenFunction.h:
```diff
@@ -226,6 +226,8 @@ class IRGenFunction {
                                  const llvm::Twine &name = "");
   void emitDeallocateDynamicAlloca(StackAddress address);
 
+  llvm::Value *emitStackAlloc(llvm::Value *size, llvm::Value *align);
```
> Can you add a comment explaining this function, specifically that it will return nullptr if the size is too large.

I don't object to comments as a rule, but the other members in this header don't have comments or indicate when they might return any given value, so I'm not certain what makes this proposed member special. Can you elaborate?

grynspan commented on lib/IRGen/GenOpaque.cpp:
```diff
@@ -629,6 +629,82 @@ void IRGenFunction::emitDeallocateDynamicAlloca(StackAddress address) {
+
+  } else {
+    // The caller specified a non-constant alignment value.
+    return nullPointer;
```
The only valid use of this builtin is to support the proposed new stdlib functions. There's no real mechanism to propagate an error there (since failures imply a fallback to the heap), so I'm not certain what the benefit of an error would be?

Are you suggesting we could emit a diagnostic message if inputs are known-bad? That's reasonable (and in an earlier pass I did have some, but I took them out pending core team feedback.)

grynspan replied:
See my reply in the other thread. :)

grynspan commented:
@swift-ci please smoke test

grynspan commented:
@swift-ci please smoke test

grynspan commented:
@swift-ci please smoke test

grynspan commented:
Rebased.

@swift-ci please smoke test

benrimmington commented on lib/AST/Builtins.cpp:
```diff
@@ -1018,6 +1018,12 @@ static ValueDecl *getDeallocOperation(ASTContext &ctx, Identifier id) {
                             _void);
 }
 
+static ValueDecl *getStackAllocOperation(ASTContext &ctx, Identifier id) {
+  return getBuiltinFunction(ctx, id, _thin,
+                            _parameters(_word, _word),
+                            _rawPointer); // FIXME: return _optional(_rawPointer)
+}
```
* [`createAsyncTaskGroupFuture`](https://github.com/apple/swift/blob/a53ab388bbbe86db37fd388201781f6a50f414c4/include/swift/AST/Builtins.def#L817) has a `Builtin.RawPointer?` parameter.

* [`getCreateAsyncTaskGroupFuture`](https://github.com/apple/swift/blob/a53ab388bbbe86db37fd388201781f6a50f414c4/lib/AST/Builtins.cpp#L1442-L1443) uses the `BuiltinFunctionBuilder` class.

```suggestion
static ValueDecl *getStackAllocOperation(ASTContext &ctx, Identifier id) {
  BuiltinFunctionBuilder builder(ctx);
  builder.addParameter(makeConcrete(ctx.getIntType())); // byteCount
  builder.addParameter(makeConcrete(ctx.getIntType())); // alignment
  builder.setResult(makeConcrete(OptionalType::get(ctx.TheRawPointerType)));
  return builder.build(id);
}
```

grynspan replied:
Thank you for the suggestion, Ben. I'm aware of those symbols. You'll note that `createAsyncTaskGroupFuture()` generates an optional value in the standard library, not in the builtin itself.

grynspan commented:
@swift-ci please smoke test

grynspan commented:
> Makes sense @jckarter. This is something I'd add as a builtin and call inside the Swift implementation (`defer { Builtin.stackDealloc(...) }`), I take it?

Went ahead and inserted the relevant lifetime-end call via a second new builtin. I'm not seeing the stack region free up—I suspect I'm still holding it wrong.

grynspan commented:
@swift-ci please smoke test

grynspan commented:
@swift-ci please smoke test

benrimmington commented on test/api-digester/stability-stdlib-abi-with-asserts.test:
```diff
@@ -56,6 +56,9 @@
 Func _collectReferencesInsideObject(_:) is a new API without @available attribute
 Func _loadDestroyTLSCounter() is a new API without @available attribute
 Func _measureRuntimeFunctionCountersDiffs(objects:_:) is a new API without @available attribute
+Func withUnsafeUninitializedMutableBufferPointer(to:capacity:_:) is a new API without @available attribute
+Func withUnsafeUninitializedMutablePointer(to:_:) is a new API without @available attribute
+Func withUnsafeUninitializedMutableRawBufferPointer(byteCount:alignment:_:) is a new API without @available attribute
```
API changes should normally be listed in the "stability-stdlib-abi-***without***-asserts.test" file instead.

https://github.com/apple/swift/blob/2950c38a5d4f1f028d3786a278ae61eaf501f0d0/test/api-digester/stability-stdlib-abi-with-asserts.test#L14-L17

However, your new APIs probably need one of:

* [`@_alwaysEmitIntoClient`](<https://github.com/apple/swift/blob/main/docs/ReferenceGuides/UnderscoredAttributes.md#_alwaysemitintoclient>) if possible; otherwise

* [`@available(macOS 9999, iOS 9999, watchOS 9999, tvOS 9999, *)`](<https://github.com/apple/swift/blob/main/docs/StandardLibraryProgrammersManual.md#availability>)

grynspan replied:
@benrimmington You're probably right—I need to check with @jckarter and/or the other core team members about the right annotation. I'm following the precedent set by other inlinable pointer functions here.

grynspan replied:
Fixed locally, pushing a change to use `@_aEIC` and remove the change to this test.

grynspan commented:
@swift-ci please test

rjmccall reviewed (dismissed):
This is using dynamic allocation and so needs to correctly interact with the `alloc_stack` stack at the SIL level.  It's okay to test this out without that, but it should not be allowed to land until that issue is fixed.

grynspan commented:
@rjmccall and I discussed off-GitHub and it seems like we'll need two IR builtins and a SIL builtin (that takes a closure), with Swift calling the SIL builtin and the SIL builtin ultimately calling the IR builtins.

lorentey commented on stdlib/public/core/UnsafeUninitializedPointer.swift:
```diff
@@ -0,0 +1,140 @@
+/// The buffer pointer passed to `body` (as well as any pointers to elements in
+/// the buffer) must not escape—it will be deallocated when `body` returns and
+/// cannot be used afterward.
+@_transparent
```
`@_transparent` still generates an ABI symbol, so as it stands these functions are lacking an `@available` declaration. We also generally prefer `@inline(__always)` in favor of `@_transparent` for new additions, unless there is a technical reason why the inlining needs to happen early on. (Is there?)

Given that the implementations are (iiuc) back-deployable, I recommend replacing `@_transparent` with:

```suggestion
@_alwaysEmitIntoClient
@inline(__always)
```

This gets rid of the superfluous ABI symbol, and still inlines these calls into clients.

(This applies to all three of the new public functions introduced here.)

xwu replied:
The proposal under review specifically states that this API will be `@_transparent`.

lorentey replied:
`@_transparent` is a non-public implementation detail that has never gone through the S-E process; I don't think the proposal should dictate such things. @grynspan I think it would make sense to remove this attribute from the API proposal.

grynspan replied:
@lorentey We can replace it (I don't think we'll need to re-run the review for that.)

grynspan commented:
> This is using dynamic allocation and so needs to correctly interact with the `alloc_stack` stack at the SIL level. It's okay to test this out without that, but it should not be allowed to land until that issue is fixed.

After investigating, `alloc_stack` and `dealloc_stack` are inappropriate here because they handle the allocation and deallocation of single stack values, not buffers/sequences of them with potentially custom alignment.

grynspan commented:
@swift-ci please test

grynspan commented:
@swift-ci please test

grynspan commented:
@swift-ci please test

grynspan commented:
@swift-ci please test

lorentey commented on stdlib/public/core/TempAlloc.swift:
```diff
@@ -0,0 +1,170 @@
+/// the buffer) must not escape—it will be deallocated when `body` returns and
+/// cannot be used afterward.
+@_alwaysEmitIntoClient @inline(__always)
+public func withUnsafeTemporaryAllocation<R>(byteCount: Int, alignment: Int, _ body: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R {
```
Formatting nit: we have an 80-character line length limit throughout the compiler & stdlib codebase.

```suggestion
public func withUnsafeTemporaryAllocation<R>(
  byteCount: Int, 
  alignment: Int, 
  _ body: (UnsafeMutableRawBufferPointer) throws -> R
  ) rethrows -> R {
```

(etc for all overlong lines)

grynspan replied:
I'll lint it all when we're sure of names. :) However I see plenty of counter-examples in the stdlib.

lorentey replied:
Mistakes happen, but nevertheless, we have a [hard 80-character line length limit](https://github.com/apple/swift/blob/main/docs/StandardLibraryProgrammersManual.md#formatting-conventions) for the stdlib. We want the stdlib to fit on regulation punch cards. 

More importantly, the 80 character limit makes it possible to read code on GitHub. Adding proper line breaks is just a basic kindness to reviewers -- it helps them understand the changes. 

These function signatures are super critical to this PR, but I currently have to awkwardly scroll around to read them, and even then, they are a pain to understand.

Compare and contrast:

```swift
@_alwaysEmitIntoClient @inline(__always)
public func withUnsafeTemporaryAllocation<R>(byteCount: Int, alignment: Int, _ body: (UnsafeMutableRawBufferPointer) throws -> R) rethrows -> R {
  ...
}
```

vs

```swift
@_alwaysEmitIntoClient @inline(__always)
public func withUnsafeTemporaryAllocation<R>(
  byteCount: Int, 
  alignment: Int, 
  _ body: (UnsafeMutableRawBufferPointer) throws -> R
) rethrows -> R {
  ...
}
```

Presentation matters.

(There is a single pragmatic exception to this rule: overlong precondition message strings are okay to leave as is. This exception most definitely doesn't apply to regular code, much less public API signatures.)

grynspan replied:
I promise I'll clean it up. :)

lorentey commented:
Nit: Commit 5f9c1027 added a `TemporaryStorageTest` binary to the root of the repo.

grynspan commented:
> Nit: Commit [5f9c102](https://github.com/apple/swift/commit/5f9c10279fe3eee469cc1643cdec336b4c267b03) added a `TemporaryStorageTest` binary to the root of the repo.

Weird, will delete.

grynspan commented:
@swift-ci please test

grynspan commented:
@swift-ci please test

grynspan commented:
@swift-ci please smoke test

grynspan commented:
@swift-ci please smoke test

xwu reviewed (commented):
Nit: Given the final name of the function, perhaps best to align the filename to `TemporaryAllocation.Swift`?

lorentey commented on stdlib/public/stubs/Stubs.cpp:
```diff
@@ -58,6 +58,10 @@
 #endif
 
+#if defined(__FreeBSD__) || defined(__OpenBSD__)
+#include <pthread_np.h>
```
Nit: this include isn't actually necessary in this file.

lorentey commented:
We still need test coverage for the new additions, including tests for expected traps. (Integer overflow, non-power-of-two alignment, etc.)

grynspan commented on stdlib/public/core/TempAlloc.swift:
```diff
@@ -0,0 +1,202 @@
+  _ body: (UnsafeMutableBufferPointer<T>) throws -> R
+) rethrows -> R {
+  _debugPrecondition(capacity > 0, "Too little uninitialized memory requested.")
+  let (byteCount, overflowed) = MemoryLayout<T>.stride.multipliedReportingOverflow(by: capacity)
+  _debugPrecondition(!overflowed, "Too much uninitialized memory requested.")
```
It is intentional that we do not trap here in release builds. This is consistent with the behaviour of existing `UnsafeMutablePointer` API that use `_debugPrecondition()`. There is a runtime cost associated with `precondition()` that cannot be optimized away if `byteCount` or `stride` is not known at compile-time. In particular, this is the case for generic callers.

grynspan commented on stdlib/public/core/TempAlloc.swift:
```diff
@@ -0,0 +1,202 @@
+  alignment: Int,
+  _ body: (UnsafeMutableRawBufferPointer) throws -> R
+) rethrows -> R {
+  _debugPrecondition(byteCount > 0, "Too little uninitialized memory requested.")
+  _debugPrecondition(alignment > 0, "Nonsensical alignment requested.")
```
See my previous reply.

For alignment in particular, after discussing with @jckarter and @rjmccall, we decided that we should silently substitute `MaximumAlignment` if the alignment value isn't known at compile-time or isn't valid, in which case the precondition doesn't prevent incorrect behaviour. It's useful for debug builds but in release builds is just runtime cost/noise.

lorentey replied:
That works great for `alignment`! 👍  

Reminder: I think we'll still need to guarantee that `Builtin.stackAlloc` will never be called with a negative byte count. (Unless that already produces a reliable trap in all builds.)

grynspan commented on stdlib/public/core/TempAlloc.swift:
```diff
@@ -0,0 +1,202 @@
+/// - Parameters:
+///   - byteCount: The number of bytes to temporarily allocate. `byteCount` must
+///     not be negative.
+///   - alignment: The alignment of the new, temporary region of allocated
+///     memory, in bytes.
```
This requirement is not stated for any of the relevant members of `UnsafeMutableRawPointer`, so it seems out of place to state it here. (It's also not true: if the alignment value is not a power of 2, we substitute `MaximumAlignment`.)

glessard replied:
The power-of-2 requirement is stated in the parameter description of the `allocate` doc-comment of UMRP (and UMRBP): https://github.com/apple/swift/blob/fb8ec6faf66ec88581f5b14b29b3f7b3811ea6f9/stdlib/public/core/UnsafeRawPointer.swift#L603
Do you think it should also be stated in the paragraph above that?

grynspan commented:
@swift-ci please test

glessard commented on stdlib/public/core/TemporaryAllocation.swift:
```diff
@@ -0,0 +1,217 @@
+/// - Parameters:
+///   - byteCount: The number of bytes to temporarily allocate. `byteCount` must
+///     not be negative.
+///   - alignment: The alignment of the new, temporary region of allocated
+///     memory, in bytes.
```
Sorry for the rehash, but if we don't state a power-of-2 requirement here, then maybe the actual behaviour should be described. The power-of-2 requirement for UMRP is stated [here](https://github.com/apple/swift/blob/fb8ec6faf66ec88581f5b14b29b3f7b3811ea6f9/stdlib/public/core/UnsafeRawPointer.swift#L603), and similarly for UMRBP.

glessard commented on stdlib/public/core/TemporaryAllocation.swift:
```diff
@@ -0,0 +1,217 @@
+/// attempts to access any such additional storage is undefined.
+///
+/// The buffer pointer passed to `body` (as well as any pointers to elements in
+/// the buffer) must not escape—it will be deallocated when `body` returns and
```
An em-dash renders too variably &mdash; might as well avoid it.

grynspan replied:
Fair enough. Curse my English degree!

grynspan commented:
@swift-ci please smoke test

lorentey commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -2969,9 +2977,76 @@ static std::unique_ptr<CallEmission> getCallEmissionForLoweredValue(
+                                             SILValue v) {
+  // The current implementation of stackAlloc() is unable to provide a
+  // stack-allocated buffer if the alignment is not known at compile-time since
+  // the emitted instructions depend on the alignment value. In practice, it is
+  // expected that almost all callers will supply a compile-time constant value,
+  // either a power of 2 or MemoryLayout<T>.stride.
```
```suggestion
  // the emitted instructions depend on the alignment value. In practice, it is
  // expected that almost all callers will supply a compile-time constant value,
  // either a direct integer literal (a power of 2) or MemoryLayout<T>.alignment.
```

grynspan commented on stdlib/public/core/TemporaryAllocation.swift:
```diff
@@ -0,0 +1,216 @@
+  }
+
+  // Slower path, with a limit known at runtime and requiring a new standard
+  // library to check.
```
FIXME: -library +library function

grynspan commented:
@swift-ci please test

grynspan commented:
@swift-ci please test linux

grynspan committed:
[SE-0322] Temporary uninitialized buffers

Adds two new IRGen-level builtins (one for allocating, the other for deallocating), a stdlib shim function for enhanced stack-promotion heuristics, and the proposed public stdlib functions.

grynspan commented:
@swift-ci please test

grynspan commented:
@swift-ci please smoke test

grynspan commented:
@swift-ci please smoke test

lorentey commented on stdlib/public/core/TemporaryAllocation.swift:
Just to double check: Is this symbol guaranteed to have weak linkage? IIUC we don't have availability for runtime entry points, and the preceding `guard #available` only protects against actually calling it; it doesn't affect linker behavior.

lorentey reviewed (approved):
Looks good from the stdlib side! 👍 

I have one question about linkage below, to ensure we won't have back deployment issues.
